### ISSUE TITLE
[BUG] `CompactProtocolFieldWriter` does not write empty value string in key-value pair

### ISSUE THREAD

**Issue by ttnghia** [rapidsai/cudf#14024]
During debugging a Parquet issue, I found that the Parquet metadata is not correctly written by cudf. In particular, in a round trip test, a pair of metadata `{"key_str", ""}` is written then read as `{"key_str", null}` in Spark. This is due to the `CompactProtocolFieldWriter` does not write empty value string:
```
size_t CompactProtocolWriter::write(KeyValue const& k)
{
  CompactProtocolFieldWriter c(*this);
  c.field_string(1, k.key);
  if (not k.value.empty()) { c.field_string(2, k.value); }                 <================ here
  return c.value();
}
```

We should write all value strings even for empty value.

**Comment by etseidl** [comment]
The `value` in `KeyValue` is an optional field in the spec, so it's correct to write nothing if it isn't set.  In #14000 I've started using `std::optional` for optional fields, to distinguish from merely empty but present fields. Perhaps that could be done throughout the Parquet thrift objects.

**Comment by vuule** [comment]
@etseidl I assume that the key-value pair is optional, but is it okay to include the key without the value?

**Comment by etseidl** [comment]
> @etseidl I assume that the key-value pair is optional, but is it okay to include the key without the value? 

According to the thrift idl, yes.

**Comment by ttnghia** [comment]
That's interesting. Then such "optional" should mean the user can specify it as an optional, instead of the current behavior (the library ignores writing empty string without the user's knowledge). So #14000 should be the way to go instead.

**Comment by etseidl** [comment]
> That's interesting. Then such "optional" should mean the user can specify it as an optional, instead of the current behavior (the library ignores writing empty string without the user's knowledge). So #14000 should be the way to go instead.

Yes, I think making more use of `optional` would make things clearer (although I didn't notice someone beat me to it :sweat_smile:  https://github.com/rapidsai/cudf/blob/ad9fa501192332ca8ce310ffe967473ec0945a97/cpp/src/io/parquet/compact_protocol_reader.hpp#L265)

**Comment by PointKernel** [comment]
Yeah, using `std::optional` to properly handle optional field reading/writing was on my TODO list but then got distracted by other tasks.

**Comment by vuule** [comment]
This issue has been at least partially addressed. We need to make another pass over `CompactProtocolReader::read` functions to verify that all optional fields are treated as such.
